content/en/blog/2023/egress-sni/index.md

```diff
@@ -0,0 +1,441 @@
+Nevertheless, the old solution was still usable with Istio versions before 1.20. Istio 1.20, however, dropped some Envoy functionality that was required for the approach to work.
+
+This post attempts to describe how we resolved the issue and filled the gap with a similar approach using Istio version-independent components and Envoy features, but without the need for a separate Nginx SNI proxy.
+Our approach allows users of the old solution to seamlessly migrate configurations before their systems face the breaking changes in Istio 1.20.
+
+## Problem to solve
+
+The currently documented egress traffic routing use-cases are all building on the assumption that the target of the traffic
+(the hostname) is statically determined by a `VirtualService`, that tells Envoy in the egress gateway pod where to TCP proxy
+the matching outbound connections. You can use multiple and even wildcard DNS names to match the routing criteria, but you
+are not able to route the traffic exactly where the application wanted to. For example you can match traffic that targets
+`*.wikipedia.org`, but you can direct everything into a single final target, i.e. `en.wikipedia.org`. If there is a
+service on `anyservice.wikipedia.org` that is not hosted by the same server(s) as `en.wikipedia.org`, your traffic to that
+arbitrary host will fail. In fact the connection would go to `en.wikipedia.org` servers, while the target hostname in the
+TLS handshake and the HTTP payload would contain `anyservice.wikipedia.org`, which may not be served by the same servers.
+The solution on high level is to inspect the original server name (SNI extension) in the application TLS handshake (that is sent
+in plain-text, so no TLS termination or other man-in-the-middle operation is needed) in every new connection and use it as
+a target dynamically to TCP proxy the traffic when leaves the gateway.
+
+## Further security measures
+
+Now that we are restricting the traffic, we should be able to lock down the egress gateways in a way, that they can be used
+only from within the mesh. This is achieved by enforcing the `ISTIO_MUTUAL` or mTLS peer authentication between the application
+sidecar and the gateway. That means, there will be two layers of TLS on the application L7 payload. One that is the application
+originated end-to-end TLS session terminated by the final remote target, and another one that is the Istio mTLS session.
+
+Also, in order to mitigate application pod corruption, not only the application sidecar performs hostname list checks, but
+the gateway as well. So that any compromised application pod can only still access the allowed targets, nothing more.
```

Comment:
```suggestion
When restricting egress traffic via egress gateways, we need to lock down the egress gateways so that they can only be used
by clients within the mesh. This is achieved by enforcing `ISTIO_MUTUAL` (mTLS peer authentication) between the application
sidecar and the gateway. That means that there will be two layers of TLS on the application L7 payload. One that is the application
originated end-to-end TLS session terminated by the final remote target, and another one that is the Istio mTLS session.

Another thing to keep in mind is that in order to mitigate any potential application pod corruption, the application sidecar and the gateway should both perform hostname list checks.
This way, any compromised application pod will still only be able to access the allowed targets and nothing more.
```